Refactor: Improve string quoting logic and fixed string-related bugs

liplum opened the pull request:
This pull request bumps the package to 2.1.0.
It doesn't have breaking changes on API level, but the end-user's output result might be changed along with the logic changes.

liplum committed:
.gitignore

- Add `.idea` to ignore IDE configuration files.

liplum committed:
feat: Improve YAML string formatting

- Add support for empty strings by using quotes.

liplum committed:
feat: Handle empty strings in YAML output

- Add a test case to verify that empty strings are correctly handled and represented as `''` in the YAML output.
- Ensure `YamlWriter` outputs `''` for an empty string value.

liplum committed:
feat: Add configuration options for YAML writing

- Introduce `YamlWriterConfig` to manage configuration.
- Allow configuration of `indentSize` and `quoteStyle` via `YamlWriterConfig`.
- Add `QuoteStyle` enum for different quote styles.
- Update tests to use `YamlWriterConfig`.
- Refactor internal code to use `YamlWriterConfig`.
- Default `quoteStyle` to `QuoteStyle.preferSingleQuote`.
- Default `indentSize` to 2.

liplum committed:
feat: Add `emptyStringLiteral` to `YamlWriterConfig`

- Add `EmptyStringLiteral` enum to represent different styles for empty strings.
- Introduce `emptyStringLiteral` in `YamlWriterConfig` with the default value `EmptyStringLiteral.singleQuote`.
- Use `context.config.emptyStringLiteral.literal` to represent the empty string in `YamlNode.toYaml` function.

liplum committed:
feat: Enhance string quoting logic and add support for unquoted strings

- Improve string quoting logic to handle single and double quotes more effectively.
  - If a string contains both single and double quotes, use a folded block scalar.
  - If a string contains only a single quote, use double quotes.
  - If a string contains only a double quote, use single quotes.
- Add `allowUnquotedStrings` option to `YamlWriterConfig` to support unquoted strings.
- Update the `QuoteStyle` enum to remove the `preferUnquoted` option.
- Introduce `literalBlockScalar` and `foldedBlockScalar` to `EmptyStringLiteral` enum.
- Add test case to check the new quoting logic and unquoted strings.
- If the text is empty, use the configured empty string literal.
- If the text is multiline, use the vertical bar.

liplum committed:
Refactor: Remove unused enum values

- Remove `literalBlockScalar` and `foldedBlockScalar` from `EmptyStringLiteral` enum in `lib/src/config.dart`.
- These values were unused.

liplum committed:
feat: Improve handling of unquoted strings in YAML

- Update `_isValidUnquotedString` to more accurately identify valid unquoted strings based on YAML specifications.
- Adjust string quoting logic based on `allowUnquotedStrings` config.
- Add more test cases to `yaml_writer_test.dart` to improve coverage and test unquoted strings in different scenarios.
- By default allowUnquotedStrings is true.

liplum committed:
feat: Add support for unquoted strings in YAML

-   Enable writing unquoted strings when `allowUnquotedStrings` is true.
-   Add tests to verify unquoted string handling with various characters and string types, such as strings with `@`, `/`, `$`, single and double quotes.
-   Add tests for different `indentSize` values (1 and 5), ensuring proper indentation in the output.
-   Add tests to check lists, empty lists and mixed data structures to improve the overall coverage.
-   Add test to check unquoted object writing.

liplum committed:
Fix: Disallow invalid characters in unquoted strings

- Prevent numbers only in unquoted string, as those will be parsed as a number.
- Add test to verify numbers only in unquoted string are quoted.
- Refactor regex to prevent numbers only in unquoted string.

liplum committed:
Update yaml writer tests

- Add tests for integers and doubles, both as raw values and stringified.

liplum committed:
feat: Handle decimal numbers in strings and custom indent sizes

- Update regex to handle decimal numbers in strings.
- Add tests for decimal numbers.
- Add support for custom indent sizes in YamlWriterConfig.
- Add more tests for indent sizes.

liplum committed:
Update package version to 3.0.0

- Bumped the package version in `pubspec.yaml` from 2.0.1 to 3.0.0.
- Updated author list in README.md.

liplum committed:
pubspec.yaml

- Update lints to ^5.1.1
- Update dependency_validator to ^5.0.2

liplum committed:
- Refactor YamlWriter constructor and library
- Make `YamlWriter` constructor `const` and add the `toEncodable` default argument.

liplum committed:
- Delete yaml_writer.iml
- Update .gitignore

liplum committed:
Refactor: Make YamlWriter constructor const

- Changed `YamlWriter` constructor to be `const`
- Updated `YamlWriterConfig` constructor to be `const`
- Updated all usages of `YamlWriter` in test and example to use `const` constructor

liplum committed:
test: Remove unnecessary print statements

- Remove unnecessary print statements in `yaml_writer_test.dart`.

liplum committed:
feat: Refactor string quoting logic in YAML writer

- Rename `allowUnquotedStrings` to `forceQuotedString` for clarity.
- Update logic to prefer unquoted strings unless `forceQuotedString` is `true`.
- Update tests to reflect the change in naming and logic.
- update comment of `forceQuotedString`

liplum committed:
Refactor: Move YamlContext class to node.dart

- Moves the `YamlContext` class definition from `yaml_context.dart` to `node.dart`.
- Deletes the now-empty `yaml_context.dart` file.
- Exports `config.dart` in `yaml_writer.dart` to expose `YamlWriterConfig`.

liplum committed:
- Remove unused import `YamlContext`
- Remove unused import `config.dart` in `yaml_writer_test.dart`
- Use `YamlWriter` instead of deprecated `YAMLWriter` type alias

gmpassos commented on lib/src/node.dart:
```diff
@@ -56,30 +60,43 @@ class StringNode extends Node {
-      !_regexpInvalidUnquotedChars.hasMatch(s) &&
-      !s.startsWith('@') &&
-      !s.startsWith('-');
+  bool _isValidUnquotedString(String s) => !_invalidCharsRegex.hasMatch(s);
```
Could you please add a unit test for this method? You may expose it as public and static.

liplum replied:
Done. I made a bunch of tests including edge cases.

gmpassos commented on lib/src/yaml_writer.dart:
```diff
@@ -12,25 +12,15 @@ typedef YAMLWriter = YamlWriter;
+  const YamlWriter({
+    this.config = const YamlWriterConfig(),
+    this.toEncodable = _defaultToEncodable,
+  });
```
How about adding a new factory constructor, `YamlWriter.config`, for these new parameters?  

This way, we can keep the unnamed constructor with the original parameters to avoid breaking the API.

liplum replied:
Done. And I bump the package version back to 2.2.0.

gmpassos reviewed (commented):
Thanks for the PR.

gmpassos commented on lib/src/node.dart:
```diff
@@ -56,30 +67,42 @@ class StringNode extends Node {
-  static final _regexpInvalidUnquotedChars = RegExp(
-      r'[^0-9a-zA-ZàèìòùÀÈÌÒÙáéíóúýÁÉÍÓÚÝâêîôûÂÊÎÔÛãñõÃÑÕäëïöüÿÄËÏÖÜŸçÇßØøÅåÆæœ@/. \t-]');
+  static final _invalidCharsRegex = RegExp(
+    r'^[:{}\[\]>,&*#?|@-]|^\s+|\s+$|\n|\t|^[0-9]+$|^[0-9]*\.[0-9]+$',
```
add `\r`

liplum replied:
Done.

liplum committed:
feat: Add default constructor for YamlWriter

- Introduce `YamlWriter()` as a new default constructor for simplified usage.
- Add `YamlWriter.config` as a named constructor for more configuration options.
- Modify all the tests to use `YamlWriter.config`.
- Update version to 2.2.0 in pubspec.yaml.
- Update README to use `YamlWriter()`.
- Update example to use `YamlWriter.config()`.

liplum committed:
feat: Improve documentation.

liplum committed:
test: Add unquoted string validation

- Add `isValidUnquotedString` method to validate unquoted strings based on YAML specifications.
- Update `isValidUnquotedString` to be a static method.
- Add test cases for `isValidUnquotedString` to cover simple text, numbers, i18n, and starting special character cases.

liplum committed:
feat: Implement comprehensive unquoted string validation in YAML

-   Implement validation rules for unquoted strings in YAML to prevent parsing errors.
-   Add a comprehensive test suite to ensure the accuracy of the validation rules.
-   Refactor the regular expression used for validation to address specific edge cases like:
    -   Invalid characters such as numbers, at signs, question marks, dashes, colons, brackets, commas, ampersands, stars, pipes, and hashes at specific positions.
    -   Invalid whitespace handling including leading, trailing, and embedded whitespace, as well as newline, tab, and carriage return characters.
-   Update tests to validate the new rules.
-   Add test cases for various characters, including edge cases like single characters, special characters, and edge cases for whitespace and hashes.

liplum committed:
Update colon tests in quote_test.dart

liplum committed:
Refactor: Update invalid character regex

- Refactor: Modify the `_invalidCharsRegex` to ensure invalid character detection.
- test: Add tests for invalid characters `#`.

liplum committed:
Refactor: Enhance YAML unquoted string validation

-   Update the `_invalidCharsRegex` to enhance the unquoted string validation logic.
-   Add new rules for invalid characters in unquoted strings:
    -   Percentage: Starting with percentage is invalid.
    -   Builtin Literal: Only a single tide is invalid, and `true`, `false`, and `null` are also invalid.
-   Add new test cases to include validation of percentages and built-in literals.
-   The regex is updated to `r"^(true|false|null|~|\?|:|-)$|^\s+|\s+$|\n|\t|\r|^[{}\[\]>,&*#|@%]|^-?(?:\d+(?:\.\d+)?|\.\d+)$|^(- |\? )|:\s|\s+#"`.
-   Update test cases to include validation of percentage and built-in literals.

liplum committed:
Refactor and enhance string handling in YAML

- Downgraded version to 2.1.0.
- Refactored string quoting for improved unquoted string handling.
- Fixed a bug where empty strings were unsupported in unquoted strings.
- Added `YamlWriter.config()` constructor and `YamlWriterConfig` class.

liplum committed:
test: Update string tests in yaml_writer_test

- Update expected test output for strings.
- Corrected string test case `s4` to `-Unquoted string.`

liplum commented:
I've added more test cases, and all tests are passed.

gmpassos commented on README.md:
```diff
@@ -75,9 +75,10 @@ Please file feature requests and bugs at the [issue tracker][tracker].
-Graciliano M. Passos: [gmpassos@GitHub][github].
+Graciliano M. Passos: [gmpassos@GitHub](https://github.com/gmpassos).
+
+Liplum: [Github](https://github.com/liplum)
```
liplum@GitHub

liplum replied:
Done

gmpassos commented on test/quote_test.dart:
```diff
@@ -0,0 +1,135 @@
+void main() {
+  group('Unquoted', () {
+    test("simple text", () {
+      expect(StringNode.isValidUnquotedString("foo"), equals(true));
```
change `equals(true)` to `isTrue`

liplum replied:
Done

gmpassos commented on test/quote_test.dart:
```diff
@@ -0,0 +1,135 @@
+
+    test("numbers", () {
+      // numbers are invalid because yaml would consider them as numbers other than strings
+      expect(StringNode.isValidUnquotedString("1"), equals(false));
```
Change `equals(false)` to `isFalse`

liplum replied:
Done

gmpassos commented:
I think that after these small adjustments, it's okay to merge.

liplum committed:
Refactor: Improve boolean assertions in quote_test.dart

- Changed assertions from `equals(true)` or `equals(false)` to `isTrue` and `isFalse` respectively, for better readability.

liplum committed:
Update GitHub link in README.md

liplum committed:
Refactor: Update string validation logic

- Add validation rules for strings starting with `>-` and `|-`.
- Update the documentation to include the new validation rule.
- Update tests to cover the new validation logic for square brackets, curly brackets, angle brackets, angle brackets + dash, pipe and pipe + dash.

liplum committed:
Doc: Handle double and single quotes in YAML strings

- When a string contains both single and double quotes, it escapes the double quote and uses double quotes for the entire string.
- When `forceQuotedString` is false and string is valid unquoted, use no quote.
- Otherwise, use the preferred quote (single or double).

liplum commented:
I think it's ready to merge.

liplum committed:
- Correct liplum@Github link in README.md.
- Disallow multi-space and colon combinations, new line and carriage return when string is unquoted in node.dart.
- Update quote tests with new restrictions.

liplum commented:
I think most edge cases are tested.

gmpassos commented on lib/src/node.dart:
```diff
@@ -56,30 +67,100 @@ class StringNode extends Node {
+        // if contains both single and double quote, escape the double quote,
+        // and use double quote
+        final result = text.replaceAll(r'\', r'\\').replaceAll('"', r'\"');
+        yamlLines.add('"$result"');
```
Use `context.config.quoteStyle` to decide which quote to use for this string case.

liplum replied:
I don't think single quote can be used for this case. Could you give me an example that single quote works in both single and double quote exisiting suitation?

gmpassos replied:
Well, this is a single-line string, so it won't escape `\r`, `\n`, or `\t`.  

It only needs to escape:  

- Double quotes (`"`): `\"` (when using `preferDoubleQuote`)  
- Single quotes (`'`): `''` (when using `preferSingleQuote`; two single quotes escape it)  

This ensures that the `QuoteStyle` is maintained.

gmpassos replied:
Do you think that we need to maintain the `QuoteStyle` in this case?

liplum replied:
Done

liplum committed:
feat: Use quote style to handle empty strings

- Refactor `YamlWriterConfig` to use `QuoteStyle` for empty strings instead of a dedicated `EmptyStringLiteral`.
- Update tests to verify empty string handling for both single and double quotes.
- Update `Node` to wrap empty string with preferred quotes.

liplum commented:
Now I use `QuoteStyle` to decide the empty string output instead of `EmptyStringLiteral` to reduce complexity.

liplum committed:
feat: Improve quote handling in YAMLWriter

- Add support for both single and double quotes in strings.
- Implement `QuoteStyle` to control whether to prefer single or double quotes.
- Update tests to cover new quoting scenarios.
- Fix handling of strings containing both single and double quotes.
- Fix strings containing single quote and prefer double quote.
- Fix strings containing double quote and prefer single quote.
- Update tests:
    - Refactor the unquoted test to use the following naming `yamlWriterUnquoted`, `yamlWriterForceQuotedSingle` and `yamlWriterForceQuotedDouble`
    - add new tests for checking strings containing both single and double quote.
    - update tests `test('list'` and `test('map'` to prefer single quote

liplum committed:
Refactor quote handling in Node class

- Modified the logic for handling strings containing both single and double quotes.
- Instead of always escaping double quotes, the code now escapes the preferred quote based on the `QuoteStyle` setting.
- Updated code to escape single quotes when `QuoteStyle` is `preferSingleQuote` and double quotes otherwise.

liplum committed:
feat: Add forceQuotedString option and improve string quoting logic

- Add `forceQuotedString` option in `YamlWriterConfig` to always quote strings.
- Refactor string quoting logic to handle cases with both single and double quotes more effectively.
- Update test cases to cover new string quoting behavior.
- If the string contains both single and double quotes, escape the preferred quote.
- If the `forceQuotedString` is false and the string is a valid unquoted string, no quotes are used.
- Otherwise, the preferred quote will be used.

liplum committed:
Refactor: Update QuoteStyle enum and improve string quoting

- Rename `QuoteStyle.preferSingleQuote` to `QuoteStyle.singleQuote`
- Rename `QuoteStyle.preferDoubleQuote` to `QuoteStyle.doubleQuote`
- Update tests to use the new `QuoteStyle` names
- Update string quoting logic for empty string to use configured quote style

gmpassos commented on lib/src/node.dart:
```diff
@@ -56,30 +68,89 @@ class StringNode extends Node {
+            break;
+          case QuoteStyle.doubleQuote:
+            result = result.replaceAll('"', r'\"');
+            yamlLines.add('"$result"');
```
With double quotes, you still need to escape any `\` character.

gmpassos replied:
Please, add a unit test for escaped characters, and if it can be read by
https://pub.dev/packages/yaml
(added as dev_dependency, only for tests).

liplum replied:
> With double quotes, you still need to escape any `\` character.

Now I escape the `\` char.

liplum replied:
> Please, add a unit test for escaped characters, and if it can be read by https://pub.dev/packages/yaml (added as dev_dependency, only for tests).

Tests are added.

liplum committed:
Refactor: Handle backslash in double quotes

- Escape backslashes before escaping double quotes when using double quote style.
- Replace single backslashes with double backslashes.

liplum committed:
feat: Add yaml escape characters support

- Add `yaml` dependency.
- Add test to check if characters like `\`, `"`, `'`, `\t` are correctly escaped.

liplum committed:
- Improve documentation
- Change default quote style to double quotes
- Change logic for unquoted strings

liplum committed:
YamlWriter: Update YamlWriter.config constructor

- Remove `const` from the `YamlWriter.config` constructor.

liplum committed:
feat: Allow configuring `QuoteStyle`

- Add `quoteStyle` config to `YamlWriterConfig`.
- Update tests to use `YamlWriter.config()` instead of `const YamlWriter.config()`.
- Change default `YamlWriter` to use `YamlWriter.config()`.
- Add `quoteStyle: QuoteStyle.singleQuote` to tests.
- Update tests for empty string and indent.
- Update `example/yaml_writer_example.dart` to use `final yamlWriter` instead of `const yamlWriter`.

gmpassos commented:
Merged.  

Thanks for your great work! I will publish it now.
